### PR TITLE
feat: add challenge editor tool

### PR DESCRIPTION
**Pull request by naomi-lgbt**
Checklist:

- [X] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [X] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [X] My pull request targets the `main` branch of freeCodeCamp.
- [X] I have tested these changes either locally on my machine, or GitPod.

Closes #XXXXX

This creates a somewhat basic tool to navigate the curriculum file tree, find the step you want to edit, and save your changes.

Once the file renaming migration is merged, will be able to test this further and mark for review.

If you want to play with this, pull down the PR, `cd` into `tools/challenge-editor`, and `npm start`.

### CODE REVIEW

**Review by ShaunSHamilton — COMMENTED** [review]
Is there a specific reason for going back to camelCase file names?

**Comment by Sembauke** [comment]
> Is there a specific reason for going back to camelCase file names?

KebabCaseIsTheWayToGo...

File: tools/challenge-editor/api/server.ts
Comment (ojeytonwilliams) [inline tools/challenge-editor/api/server.ts:47]:
```suggestion
const app = express();

app.use(
  cors({
    origin: 'http://localhost:3300'
  })
);

app.use(express.static('public'));
app.use(express.json());
app.use(express.urlencoded());

app.post('/:superblock/:block/_tools/:command', (req, res, next) => {
  toolsRoute(req, res).catch(next);
});

app.post('/:superblock/:block/:step', (req, res, next) => {
  saveRoute(req, res).catch(next);
});

app.get('/:superblock/:block/:step', (req, res, next) => {
  stepRoute(req, res).then(next).catch(next);
});

app.get('/:superblock/:block', (req, res, next) => {
  blockRoute(req, res).catch(next);
});

app.get('/:superblock', (req, res, next) => {
  superblockRoute(req, res).catch(next);
});

app.get('/', indexRoute);

app.listen(3200, () => console.log('App is live on 3200!'));
```
Since this is all synchronous at the top level, there's no need to wrap this in an IIFE.  Also, `async` functions aren't supported by express 4 (apparently they're coming in 5).  It's not a disaster to use them, but any custom error handling won't work.

File: tools/challenge-editor/api/server.ts
Comment (ojeytonwilliams) [inline tools/challenge-editor/api/server.ts:1]:
```suggestion
```
Not strictly necessary. See below.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Hey Naomi, I have a few ideas how to make this a little cleaner and more standard.  Let me know what you think.

File: tools/challenge-editor/api/routes/toolsRoute.ts
Comment (ShaunSHamilton) [reply tools/challenge-editor/api/routes/toolsRoute.ts:66]: ENUM!

Comment (ojeytonwilliams) [reply tools/challenge-editor/api/routes/toolsRoute.ts:66]: I was thinking of you when I was looking at this :D

Comment (nhcarrigan) [reply tools/challenge-editor/api/routes/toolsRoute.ts:66]: Well now I can't do it. 😉

File: tools/challenge-editor/client/src/components/editor/Editor.tsx
Comment (ojeytonwilliams) [inline tools/challenge-editor/client/src/components/editor/Editor.tsx:44]:
```suggestion
        (error: Error) => {
          setLoading(false);
          setError(error);
        }
```

File: tools/challenge-editor/client/src/components/editor/Editor.tsx
Comment (ojeytonwilliams) [inline tools/challenge-editor/client/src/components/editor/Editor.tsx:48]:
```suggestion
  const handleChange = (
```
`changeHandle` is a bit of a Yodaism.

File: tools/challenge-editor/client/src/components/editor/Editor.tsx
Comment (ojeytonwilliams) [inline tools/challenge-editor/client/src/components/editor/Editor.tsx:67]:
```suggestion
        onBeforeChange={handleChange}
```

File: tools/challenge-editor/client/src/components/editor/Editor.tsx
Comment (ojeytonwilliams) [inline tools/challenge-editor/client/src/components/editor/Editor.tsx:57]:
```suggestion
    return <div>Error: {error.message}</div>;
```

File: tools/challenge-editor/client/src/components/landing/Landing.tsx
Comment (ojeytonwilliams) [inline tools/challenge-editor/client/src/components/landing/Landing.tsx:6]:
```suggestion
  const [error, setError] = useState<Error | null>(null);
```

File: tools/challenge-editor/client/src/components/landing/Landing.tsx
Comment (ojeytonwilliams) [inline tools/challenge-editor/client/src/components/landing/Landing.tsx:27]:
```suggestion
        (error: Error) => {
          setLoading(false);
          setError(error);
        }
```

File: tools/challenge-editor/client/src/components/landing/Landing.tsx
Comment (ojeytonwilliams) [inline tools/challenge-editor/client/src/components/landing/Landing.tsx:32]:
```suggestion
    return <div>Error: {error.message}</div>;
```

File: tools/challenge-editor/client/src/components/superblock/SuperBlock.tsx
Comment (ojeytonwilliams) [inline tools/challenge-editor/client/src/components/superblock/SuperBlock.tsx:6]:
```suggestion
  const [error, setError] = useState<Error | null>(null);
```

File: tools/challenge-editor/client/src/components/superblock/SuperBlock.tsx
Comment (ojeytonwilliams) [inline tools/challenge-editor/client/src/components/superblock/SuperBlock.tsx:29]:
```suggestion
        (error: Error) => {
          setLoading(false);
          setError(error);
        }
```

File: tools/challenge-editor/client/src/components/superblock/SuperBlock.tsx
Comment (ojeytonwilliams) [inline tools/challenge-editor/client/src/components/superblock/SuperBlock.tsx:34]:
```suggestion
    return <div>Error: {error.message}</div>;
```

File: tools/challenge-editor/client/src/reportWebVitals.ts
Comment (ojeytonwilliams) [inline tools/challenge-editor/client/src/reportWebVitals.ts:17]: Do we need/want this for a local tool?

File: tools/challenge-editor/package.json
Comment (ojeytonwilliams) [inline tools/challenge-editor/package.json:13]: Since this just holds some scripts, I think it would be better to pull those scripts to the root and delete this.  Then contributors don't need to go hunting for this.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Review part two:

File: tools/challenge-editor/api/server.ts
Comment (ojeytonwilliams) [inline tools/challenge-editor/api/server.ts:32]:
```suggestion
app.get('/:superblock/:block/:step', (req, res, next) => {
  stepRoute(req, res).catch(next);
});
```
Sorry, just randomly spotted this :/

**Comment by moT01** [comment]
I tried running this last week or something and couldn't quite get it working. Is there any other steps I'm missing that I need to do to get it running @nhcarrigan?

**Comment by nhcarrigan** [comment]
I just pulled it down again and it came right up for me. You should be in the root directory and run `npm run challenge-editor`. Might need to do an `npm ci` first as I've added a couple of deps?

**Review by moT01 — DISMISSED** [review]
This looks real good @nhcarrigan 🎉 I would be fine merging it like this.

The only thing I thought we could maybe address on this PR is that there's some leftover create-react-app assets and things - the react logo file, and the content of the README, might be some other stuff. Could probably just delete those.

#### Other observations:
At the bottom of the superblock pages it says, "Want to create a new project? Open your terminal, point to the root directory for this repo, and run `npm run create-project`." That doesn’t seem to work. Need to go into the challenge helpers folder I suppose.

Wondering what content we should include in the map here. Maybe we should omit the cert projects since there's just one file? Maybe we should just include all the superblocks? Doesn't really matter for now, just throwing ideas out there. I would lean towards including all the superblocks - people may start to prefer using this editor.

#### UI thoughts:
Save changes button doesn't give any feedback or 'seem' like it does anything. It saves the file, but I just have a regular mouse when hovering and clicking doesn't give any visual feedback. Similar with the buttons on the step-tools page - although, I do get an alert with those.

The editor is kind of small - would like to see it take as much of the screen as possible.

There's no syntax highlighting for javascript - there is for html and css.

I don't know what file I am editing. I just see `step-X`. I guess I can see the superblock and block in the address bar, could maybe add those to the UI.

Could maybe add some buttons to easily move between steps, similar to how we can move between steps on /learn with the hotkeys.

On the step-tools page, there's no link back to the block page - can't easily get back there. I suppose I can with the back button.

Might like to see the list of steps on the step-tools page. Could maybe just move the step-tools page to the bottom of the block page so you can see that list.

<hr>

These are mostly just observations that I think might improve things. I don't think we need to hold off on merging this because of any of them. It works real good and we can continue to improve it down the road.

**Comment by nhcarrigan** [comment]
> At the bottom of the superblock pages it says, "Want to create a new project? Open your terminal, point to the root directory for this repo, and run npm run create-project." That doesn’t seem to work. Need to go into the challenge helpers folder I suppose.

Updated!

> I would lean towards including all the superblocks - people may start to prefer using this editor.

I went ahead and added all the superblocks. We can revert this if we change our mind.

> Save changes button doesn't give any feedback or 'seem' like it does anything. It saves the file, but I just have a regular mouse when hovering and clicking doesn't give any visual feedback. Similar with the buttons on the step-tools page - although, I do get an alert with those.

Updated the CSS for the active button event, so that should be more clear. You were supposed to be getting an alert for the save button too, but I messed up the response format. That's been fixed now.

> There's no syntax highlighting for javascript - there is for html and css.

Imported JavaScript and JSX highlighting.

> I don't know what file I am editing. I just see step-X. I guess I can see the superblock and block in the address bar, could maybe add those to the UI.

Added "breadcrumbs" (they aren't clickable) that indicates superblock and block.

> On the step-tools page, there's no link back to the block page - can't easily get back there. I suppose I can with the back button.

Added a link.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Hey @nhcarrigan I went ahead and cleaned out some old deps, since I was poking around.  I also found some pretty minor issues, but it's looking great.

**Review by ojeytonwilliams — DISMISSED** [review]
LGTM :+1:

**Review by moT01 — CHANGES_REQUESTED** [review]
There's no syntax highlighting for pug...
Screen Shot 2022-04-15 at 10 54 30 AM

**Review by moT01 — DISMISSED** [review]
jk - LGTM 🎉 (after tests pass)

**Comment by nhcarrigan** [comment]
@moT01 @ojeytonwilliams Rebase complete.

**Review by ojeytonwilliams — APPROVED** [review]
:+1: :+1:

**Review by moT01 — APPROVED** [review]

**Pull request merged**
